src/e3/anod/qualifiers_manager.py

```diff
@@ -219,6 +227,122 @@ def repr(self, value: str | bool, hash_pool: list[str] | None) -> str:
+            (used in error messages)
+        :param name: see QualifierDeclaration.__init__
+        :param description: see QualifierDeclaration.__init__
+        :param repr_in_hash: see QualifierDeclaration.__init__
+        :param repr_name: see QualifierDeclaration.__init__
+        :param repr_omit_key: if True discard qualifier name in string representation
+        :param default: default value for the qualifier
+        :param choices: list of valid value for the qualifier
+        """
+        super().__init__(
+            origin=origin,
+            name=name,
+            description=description,
+            repr_in_hash=repr_in_hash,
+            repr_name=repr_name,
+        )
+        self.repr_omit_key = repr_omit_key
+
+        # Check if the default is valid
+        default_values: frozenset[str] | None = None
+        if default is not None:
+            default_values = frozenset(default.split(";"))
+            if choices is not None:
+                wrong_values = default_values - set(choices)
+
+                if wrong_values:
+                    choices_str = ", ".join((f"'{choice}'" for choice in choices))
+                    wrong_values_str = ", ".join(
+                        (f"'{value}'" for value in sorted(wrong_values))
+                    )
+                    raise AnodError(
+                        f"{self.origin}: In '{self.name}', default value(s) "
+                        f"({wrong_values_str}) should be in ({choices_str})"
+                    )
+
+        self.choices = choices
+        self._default = default_values
+
+    @property
+    def default(self) -> frozenset[str] | None:
+        """See QualifierDeclaration.default."""
+        return self._default
+
+    def value(self, value: str | bool | frozenset[str]) -> frozenset[str]:
+        """See QualifierDeclaration.value."""
+        assert isinstance(value, str)
+
+        # Make sure '' value is the empty set
+        value_set = frozenset(value.split(";")) if value else frozenset({})
+
+        # Check if the values are in choices
+        if self.choices:
+            wrong_values = value_set - set(self.choices)
+
+            if wrong_values:
+                choices_str = ", ".join((f"'{choice}'" for choice in self.choices))
+                wrong_values_str = ", ".join(
+                    (f"'{value}'" for value in sorted(wrong_values))
+                )
+                raise AnodError(
+                    f"{self.origin}: Invalid value(s) for qualifier {self.name}: "
+                    f"({wrong_values_str}) not in ({choices_str})"
+                )
+
+        return value_set
+
+    def repr(
+        self, value: str | bool | frozenset[str], hash_pool: list[str] | None
+    ) -> str:
+        """See QualifierDeclaration.repr."""
+        assert isinstance(value, frozenset)
+        if not value:
+            # An empty value for tag value should lead to an empty representation
```

Comment:
it seems that the comment is not corresponding the key_set case here

Reply:
Done